### PR TITLE
bpo-40334: produce specialised errors for del

### PR DESCRIPTION
**Pull request by hauntsaninja**
https://bugs.python.org/issue40334

### CODE REVIEW

**Comment by hauntsaninja** [comment]
I hope it's okay that I reused the main BPO for this? If not, let me know, and I'll file and link against separate bugs in the future.
This would check a todo off in https://github.com/we-like-parsers/cpython/issues/122

**Review by gvanrossum — COMMENTED** [review]
It seems there aren't enough tests for this. :-(

With the old parser I get sensible errors ("cannot delete function call" / "operator") for things like `del a, b()`, `del a(), b` and `del a+b`. With the new parser and this PR those get weird errors:
```
>>> del a, b()
  File "<stdin>", line 1
    del a, b()
              ^
SyntaxError: invalid syntax
>>> del a(), b
  File "<stdin>", line 1
    del a(), b
    ^
SyntaxError: cannot delete tuple
>>> del a+b
  File "<stdin>", line 1
    del a+b
         ^
SyntaxError: invalid syntax
>>> 
```

**Comment by hauntsaninja** [comment]
Thanks! I just found and fixed the multiple delete function call cases, but I'll look into the addition one. I'll convert this to a draft PR for now.

**Comment by hauntsaninja** [comment]
Here, it should be better now.
The column offsets are still a little misplaced, since they point to the end of the expression. I might just be missing a way to fix this, can also change it to use `RAISE_SYNTAX_ERROR_NO_COL_OFFSET`.

**Comment by lysnikolaou** [comment]
The doctest on https://github.com/python/cpython/blob/f01d1be97d740ea0369379ca305646a26694236e/Lib/test/test_syntax.py#L67 also need to be unskipped.

**Comment by lysnikolaou** [comment]
@pablogsal Should we maybe close and reopen this for the Github Actions to run as well?

**Comment by pablogsal** [comment]
> @pablogsal @gvanrossum Should we maybe close and reopen this for the Github Actions to run as well?

**Comment by lysnikolaou** [comment]
> Is odd that GH actions didn't execute....

Probably some network hiccup.

File: Grammar/python.gram
Comment (lysnikolaou) [inline Grammar/python.gram]: I think that this deserves a comment which explains why the lookahead's needed.

**Comment by hauntsaninja** [comment]
Thank you for the notes and for getting CI working!

File: Grammar/python.gram
Comment (pablogsal) [reply Grammar/python.gram]: This won't work for something like:

```
del a.b.c + 2
```

The old parser correctly points out:

```
>>> del a.b.c + 2
  File "<console>", line 1
SyntaxError: cannot delete operator
```

but with this PR:

```
>>> del a.b.c + 2
  File "<console>", line 1
    del a.b.c + 2
              ^
SyntaxError: invalid syntax
```

Comment (pablogsal) [reply Grammar/python.gram]: Also, there is some other differences. For instance where the old parser says:

```
>>> del (a, b, (e, d.e.f+2))
  File "<console>", line 1
SyntaxError: cannot delete operator
```

the new parser (with this PR) says:

```
>>> del (a, b, (e, d.e.f+2))
  File "<console>", line 1
    del (a, b, (e, d.e.f+2))
                           ^
SyntaxError: cannot delete tuple
```

**Comment by hauntsaninja** [comment]
Thanks! Fixed those test cases, and added some others.

I hope this PR is helpful; if this is a situation where it's easy for you to see the fix and roundtripping things with me is just slowing things down, I'm genuinely happy to close this in the pursuit of efficient use of core dev time :-)

**Comment by lysnikolaou** [comment]
> I hope this PR is helpful

It is and thanks for taking the time to work on this.

Although it is helpful, I feel kind of hesitant to approve this, because I fear that we might be missing something in the `del_targets_end` rule. To be completely honest it is also a bit ugly, though necessary at the moment. I think I'd be okay to merge this and hopefully we will be able to get rid of the ugly lookahead situation, if we implement labeled exceptions. I'm leaving the final decision up to @gvanrossum and @pablogsal.

File: Grammar/python.gram
Comment (gvanrossum) [inline Grammar/python.gram:598]: Shouldn't this also include `ENDMARKER`? (Though I don't know how to exercise that.)

**Review by gvanrossum — COMMENTED** [review]
I am hesitant too, and the asymmetry between `del_target` and other types targets worries me.

At the same time I have a feeling that implementing the throw/catch mechanism is a very large change that we might want to put off till 3.10. (I certainly don't want to rush it in before beta1, but I could be talked into aiming for a later beta -- the last one goes out July 20th.)

**Comment by pablogsal** [comment]
> I hope this PR is helpful;

Thanks for working on this @hauntsaninja!

I feel very similar to @lysnikolaou regarding the targets. Although I think there is not a clear way to do this thing better than the current proposal, I still feel that is somehow working around some fundamental limitation that we have. Especially since we are starting to alter other rules in order to raise specific syntax errors in different places. What worries me the most about this is the maintainability of this: things can start feeling "magical" and not super evident (on top of the particularities of PEG).

One thing we could do is leave this open until we gather some experience with labelled failures or at least think more about error reporting and merge it if we don't have anything before the first release candidate.

**Comment by lysnikolaou** [comment]
> Shouldn't this also include ENDMARKER?

Nope, I think there currently isn't a way for an `ENDMARKER` to immediately follow a `del_stmt` without an intermediate `NEWLINE`.

**Comment by hauntsaninja** [comment]
Agreed that this feels hacky. I've spent less time thinking about parsing than you all have, nor have I gone through the labelled exceptions proposal. So don't yet have much to contribute to this decision and very happy / it makes sense to leave this open in search of a better solution.

If we do need to merge this in, just want to note that the column offset situation is also not currently ideal, since we point past the end of the un-del-able expression, rather than at its beginning. We could always use `RAISE_SYNTAX_ERROR_NO_COL_OFFSET` though.

I added in ENDMARKER, although Lysandros is right that we don't need it, since we always parse a NEWLINE in simple_stmt.

**Comment by pablogsal** [comment]
> just want to note that the column offset situation is also not currently ideal, since we point past the end of the un-del-able expression

Yeah, I think actually this can be pretty confusing to the user. I think the old parser didn't report the column offset, no?

**Comment by lysnikolaou** [comment]
> > just want to note that the column offset situation is also not currently ideal, since we point past the end of the un-del-able expression
> 
> Yeah, I think actually this can be pretty confusing to the user. I think the old parser didn't report the column offset, no?

Correct.

**Comment by hauntsaninja** [comment]
Huh, I use IPython a lot, so didn't quite realise these SyntaxError's didn't report the column offset to the user. Still, `e.offset` did get set on the error, mostly at the place you'd want.

<details>

```
λ python3 -m IPython
Python 3.8.2 (default, Apr 21 2020, 00:39:48) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: del x, y, z, f(), a, b, c                                                                                                                      
  File "<ipython-input-1-2e07c1069205>", line 1
    del x, y, z, f(), a, b, c
                 ^
SyntaxError: cannot delete function call

λ cat test73.py
try:
    s = input()
    compile(s, "<>", "exec")
except SyntaxError as e:
    print(e)
    print(e.offset)
    print(s)
    print(" " * (e.offset - 1) + "^")

λ python3 test73.py
del a, b + c
cannot delete operator (<>, line 1)
8
del a, b + c
       ^

λ python3 test73.py
del x, y, z, f(), a, b, c
cannot delete function call (<>, line 1)
14
del x, y, z, f(), a, b, c
             ^

λ python3 test73.py
del [a, b, (c, d.e.f[0] + 2)]
cannot delete operator (<>, line 1)
5
del [a, b, (c, d.e.f[0] + 2)]
    ^
```
</details>

**Review by gvanrossum — COMMENTED** [review]
Okay, I have thought about this some more, and this is actually pretty good. I'd say remove the ENDMARKER (having looked at the grammar for `simple_stmt` I agree it's pointless), and resolve the conflict by regenerating parse.c, and then I'll approve and land it.

Sorry this took so long to ponder!

It's actually better than what we do for assignments, e.g. compare `del f(), g` to `f(), g = 1`.

What's weird is that the old parser puts the caret at the *start* of the offending expression, while the new parser puts it at the *end*. That's harder to fix, and I propose not to bother.

**Comment by hauntsaninja** [comment]
ENDMARKER removed and pegen has been regen-ed!

**Review by gvanrossum — APPROVED** [review]
Excellent! Thanks so much for helping out here.

**Comment by lysnikolaou** [comment]
Thanks @hauntsaninja.

**Pull request merged**
